docs/proposals/maintenance.md

```diff
@@ -0,0 +1,179 @@
+
+### One-shot drain of multiple nodes
+
+In this case, we try to meet some specific service goals specified by the admin
+on the command line.  The specification of these parameters only when we run the
+tool lets us provide the possibility of safety while avoiding questions of where
+to configure and store such specification (treated further down).
+
+```
+kubectl drain \
+  --min-shard-strength=0.75 \
+  --min-seconds-between-evictions=900 \
+  --grace=900 $nodes
+```
+
+This would still be pretty simple.  We offer no guarantees about safety in the
+face of crashes or restarts, but we do meet the needs of admins who want
+something more sophisticated than `sleep`.
+
+### Implementation of the above as a controller
+
+The parameters of a specific lame, drain, or unlame request could be pushed to
+the API server as a drain object.  The drain controller would have to manage
+multiple drains in flight at once.  If the drains have differing parameters,
+then more aggressive drains would generally starve safer ones.  While this could
+be a useful feature in emergencies, it will be a lot easier to reason about
+multiple drains in progress if they all share the same parameters, which would
+be configured via kube-env and would represent the cell admin's SLO.
+
+The drain object would thus look something like this
+
+```
+type DrainSpec struct {
+  LameNodes    PodList
+  RemoveNodes  PodList
+  AddNodes     PodList
+  GracePeriod  int
+}
+```
+
+A drain object would work much like a deployment.  A drain controller would
+watch for drain objects and, within constraints, work toward the desired end
+state.  Nodes in `LameNodes` would merely be lamed, nodes is `RemoveNodes` would
+be drained according to the given parameters, and nodes in `AddNodes` would be
+undrained.  The object would be rejected if the sets overlap at all.  The action
+of the drain controller can be paused or stopped by deleting the drain object,
+and restarted by re-posting it.
+
+Undo would be trickier, but a clever client could be written such that it
+records the starting state of all the nodes and is able to send a reversing
+drain request.
+
+A central drain controller would have several advantages: safe restarts and the
+ability to react to other events in order to preserve the desired SLO.  For
+example, if a machine not in the list of RemoveNodes crashes and takes out other
+pods in replication sets also impacted by the drain, it can hold off on
+evictions impacting that replication set until MinSecondsBetweenEvictions and
+MinShardStrength are OK again.
+
+## Implementation of safety logic
+
+The implementation is the same whether there is one drain in flight or several.
+The drain agent (either kubectl or the drain controller) must watch for
+replication controllers and pods, and it will watch all pods to notice any time
+there is a disruption to a ReplicaSet.  Because there is no reliable mechanism
```

Comment:
I think you can do this without watching for disruptions.  It can just watch the current state: watch all RCs, check that each RC has as at least `drain.minShardStrength * rc.spec.replicas` pods, all of which are at least `minSecondsBetweenEvictions` old.